include/picongpu/plugins/saxs/beam/CoordinateTransform.hpp

```diff
@@ -0,0 +1,169 @@
+/* Copyright 2020 Pawel Ordyna
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+#include "picongpu/plugins/saxs/beam/Side.hpp"
+#include "picongpu/plugins/saxs/beam/SecondaryRotation.hpp"
+#include "picongpu/param/saxs.param"
+
+namespace picongpu
+{
+namespace plugins
+{
+namespace saxs
+{
+namespace beam
+{
+    //! Get the global domain size as a 3D vector in 3D and 2D simulations.
+    template < unsigned DIM >
+    HINLINE float3_X getDomainSize ( );
+
+    // For 3D simulations:
+    template < >
+    HINLINE float3_X getDomainSize< DIM3 > ( )
+    {
+        DataSpace< DIM3 >  globalDomainSize = Environment< DIM3 >::get(
+            ).SubGrid( ).getGlobalDomain( ).size;
+        return precisionCast< float_X >( globalDomainSize );
+    }
+
+    // For 2D simulations:
+    template< >
+    HINLINE float3_X getDomainSize< DIM2 > ( )
+    {
+        auto  globalDomainSize = Environment< DIM2 >::get( ).SubGrid(
+            ).getGlobalDomain().size;
+        return float3_X(
+            globalDomainSize[ 0 ],
+            globalDomainSize[ 1 ],
+            0.0
+        );
```

Comment:
(Not important)
This does not have to be a template. Alternatively, you could just have a single non-template function, create a `float3_X` of zeroes, then `auto  globalDomainSize = Environment< simDim >::get( ).SubGrid().getGlobalDomain().size` and make a look for `simDim` iterations copying from `globalDomainSize` to your result.